File: Stack/C/Stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int data;
    struct node* next;
} node;

node* create_node() {
    node* new = (node*) malloc(sizeof(node));
    return new;
}

void push(node** stack, int data) {
    node* new = create_node();

    if((*stack) == NULL) {
        new->data = data;
        new->next = NULL;

        (*stack) = new;
        return;
    }

    new->data = data;
    new->next = (*stack);
    (*stack) = new;

    return;
}

void push_array(node** stack, int* data, int size) {
    for (register int i = 0; i < size; i++) {
        push(&stack, data[i]);
    }
}

void push_node(node** stack, node* data) {
    node* aux = data;

    while(aux != NULL) {
        push(&stack, aux->data);
        aux = aux->next;
    }
}

int pop(node** stack) {
    if((*stack) == NULL) {
        return NULL;
    }

    node* aux = (*stack);
    int value = (*stack)->data;
    
    (*stack) = (*stack)->next;
    free(aux);

    return value;
}

void print(node* stack) {
    node* aux = stack;

    while(aux != NULL) {
        printf("%d\n", aux->data);
        aux = aux->next;
    }
    printf("\n");
}

int peak(node* stack) {
    return stack->data;
}

bool is_empty(node *stack) {
    if(stack == NULL) {
        return true;
    }
    return false;
}

int length(node* stack) {
    node* aux = stack;
    int counter = 0;

    while(aux != NULL) {
        counter++;
        aux = aux->next;
    }

    return counter;
}

int last(node* stack) {
    return peak(stack);
}

int get_value_by_index(node* stack, int index) {
    node* new_stack = NULL;
    int aux, value, counter = 0;
    bool found = false;

    while(stack != NULL) {
        push(&new_stack, pop(&stack));
    }

    while(new_stack != NULL) {
        aux = pop(&new_stack);
        push(&stack, aux);        
        
        if(counter == index) {
            value = aux;
            found = true;
        }
        
        counter++;
    }

    if(found == false) {
        return -1;
    }

    return value;
}

int get_index_by_value(node* stack, int value) {
    node* new_stack = NULL;
    int aux, index, counter = 0;
    bool found = false;

    while(stack != NULL) {
        push(&new_stack, pop(&stack));
    }

    while(new_stack != NULL) {
        aux = pop(&new_stack);
        push(&stack, aux);

        if(aux == value) {
            found = true;
            index = counter;
        }
        counter++;
    }

    if(found == false) {
        return -1;
    }

    return index;
}

int* get_all_indexes_by_value(node* stack, int value) {
    node* new_stack = NULL;
    int* indexes = NULL;
    int aux, counter = 0, array_counter = 0;

    while(stack != NULL) {
        push(&new_stack, pop(&stack));
    }

    while(new_stack != NULL) {
        aux = pop(&new_stack);

        if(aux == value) {
            indexes = malloc(sizeof(int));
            *(indexes + array_counter) = aux;
            array_counter++;
        }

        counter++;
    }

    return indexes;
}

int main() {
    node *stack = NULL;

    push(&stack, 1);
    push(&stack, 2);
    push(&stack, 3);
    push(&stack, 4);
    push(&stack, 5);
    print(stack);
    pop(&stack);
    printf("valor = %d\n", get_value_by_index(stack, 5));
    print(stack);
    return 0;
}

File: Stack/Stack.ts
class Stack { // peak = 0, base = n
    data: any = null
    next: Stack | null = null

    constructor(data: any, next: Stack | null) {
        this.data = data
        this.next = next
    }

    push(data: any) {
        if(data === null) {
            return
        }

        if(this.data === null) {
            this.data = data
            return
        }

        this.next = new Stack(this.data, this.next)
        this.data = data
        return
    }

    pop() {
        if(this.data === null) {
            return null
        }

        if(this.next === null) {
            let data: any = this.data
            this.data = null
            
            return data
        }

        let data: any = this.data

        this.data = this.next.data
        this.next = this.next.next

        return data
    }

    peak() {
        return this.data
    }

    isEmpty() {
        return this.data === null
    }

    length() {
        let aux: Stack = new Stack(this.data, this.next)
        let counter: number = 0;

        while(aux.next !== null) {
            counter = counter + 1
            aux = aux.next
        }

        return counter + 1
    }

    last() {
        let aux: Stack = new Stack(this.data, this.next)

        while(aux.next !== null) {
            aux = aux.next
        }

        return aux.data
    }

    getValueByIndex(index: number) { 
        let aux: Stack = new Stack(this.data, this.next)
        
        for(let i = 0; i < aux.length(); i++) {
            if(i === index) {
                return aux.data
            }

            aux = aux.next
        }

        return null
    }

    getIndexByValue(value: any) {
        let aux: Stack = new Stack(this.data, this.next)
        
        for(let i = 0; i < aux.length(); i++) {
            if(aux.data === value) {
                return i
            }

            aux = aux.next
        }

        return null
    }

    getAllIndexesByValue(value: any) {
        let aux: Stack = new Stack(this.data, this.next)
        let indexes: Array<number> = []

        for(let i = 0; i < aux.length(); i++) {
            if(aux.data === value) {
                indexes.push(i)
            }

            aux = aux.next
        }

        return indexes
    }

    getAllValuesByIndexes(indexes: Array<number>) {
        let aux: Stack = new Stack(this.data, this.next)
        let values: Array<any> = []

        for(let i = 0; i < aux.length(); i++) {
            if(i in indexes) {
                values.push(aux.data)
            }

            aux = aux.next
        }

        return values
    }

    getValuesBySlice(begin: number, end: number) {
        let aux: Stack = new Stack(this.data, this.next)
        let values: Array<any> = []

        for(let i = 0; i < aux.length(); i++) {
            if(begin <= i && i <= end) {
                values.push(aux.data)
            }

            aux = aux.next
        }

        return values
    }

    removeAll() {
        this.data = null
        this.next = null
    }

    removeByIndex(index: number) {
        let aux: Stack = new Stack(this.data, this.next)
    }
}
